### PR TITLE
Introduce HTTP OTLP exporter

### LINKED ISSUE #21535: Provide OpenTelemetry OTLP/HTTP exporter
### Description

The current `quarkus-opentelemetry-exporter-otlp` extension exports OpenTelemetry traces using the OTLP/gRPC protocol (using the `OtlpGrpcSpanExporter` from the `io.opentelemetry:opentelemetry-exporter-otlp-trace` module). The OpenTelemetry SDK however also includes a `OtlpHttpSpanExporter` (in module `io.opentelemetry:opentelemetry-exporter-otlp-http-trace`).

We would like to be able to export our OpenTelemetry traces using the OTLP/HTTP transport (see also https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#otlphttp).

### Implementation ideas

I think the most straight-forward solution would be to provide a new extension with a config namespace of `quarkus.opentelemetry.tracer.exporter.otlp-http` (in contrast to the current `quarkus.opentelemetry.tracer.exporter.otlp` which would pertain to the OTLP/gRPC exporter only). The config keys would basically be exactly the same, as the OTLP/HTTP exporter builder exposes the same configuration options.

The user would then have choice of either adding the `quarkus-opentelemetry-exporter-otlp` or the `quarkus-opentelemetry-exporter-otlp-http` extension.

### PR DESCRIPTION
**Pull request by geoand**
The integration utilizes the `HttpSender`
interface introduced in OTel 1.28
and is based on the Vert.x HTTP Client

Closes: #21535

### CODE REVIEW

**Comment by geoand** [comment]
@yrodiere I am seeing:

```
Error: Class initialization of org.h2.fulltext.FullTextLucene failed. This error is reported at image build time because class org.h2.fulltext.FullTextLucene is registered for linking at image build time by command line Use the option --initialize-at-run-time=org.h2.fulltext.FullTextLucene to explicitly request delayed initialization of this class.
Original exception that caused the problem: java.lang.NoClassDefFoundError: org/apache/lucene/index/IndexFormatTooOldException
```

in the main integration test and it should be completely unrelated to this PR (or at least I don't see how it could be related)

**Comment by yrodiere** [comment]
I'm back from two weeks out of office, so take this with a grain of salt, but all I can say is that I didn't see any change related to H2, nor did I ever see any use of Lucene in Quarkus Core (let alone H2).

I suppose your changes might be introducing code that makes GraalVM think that `FullTextLucene` is reachable, even if it's not.

But be aware there is (used to be?) a bug in GraalVM in which the firsts static init error was ignored and GraalVM still proceeds and generates tons of seemingly nonsensical errors as a result. See https://github.com/oracle/graal/issues/6368

I can have a look at some point, but I hope it's not too urgent as I'm still catching up with emails and (lots of) PRs that I need to review to get others unstuck.

**Comment by yrodiere** [comment]
Ah wait, I'm seeing this elsewhere, so it's not limited to this PR.

I guess I know what I'll do with my day :)

**Comment by yrodiere** [comment]
The caused seems to be there: https://github.com/quarkusio/quarkus/pull/34615#issuecomment-1661613280

**Comment by geoand** [comment]
Thanks for the super quick investigation!

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java [hunk 1]
```diff
@@ -89,9 +91,9 @@ interface TrustCert {
         Optional<List<String>> certs();
     }
 
-    public static class Protocol {
+    class Protocol {
         public static final String GRPC = "grpc";
-        public static final String HTTP_PROTOBUF = "http/protobuf";
-        public static final String HTTP_JSON = "http/json";
+        public static final String HTTP_PROTOBUF = "http";
```
Comment (brunobat) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java:96]: The correct value is "http/protobuf", per spec.
See: https://github.com/open-telemetry/opentelemetry-java/blob/7ae32c90e012d87000bd3f64b9c4b3b937f85de9/exporters/otlp/all/src/main/java/io/opentelemetry/exporter/otlp/internal/OtlpConfigUtil.java#L42

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java [hunk 2]
```diff
@@ -89,9 +91,9 @@ interface TrustCert {
...
+        public static final String HTTP_JSON = "http+json";
```
Comment (brunobat) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java:97]: The value for HTTP_JSON is not defined yet, we should not create one that most likely will change.
The current JSON support is for local serialisation and storage, prior to send to the collector.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxHttpExporter.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                                        onError.accept(callResult.cause());
+                                    }
+                                }
+                            })
+                                    .putHeader("Content-Type", contentType);
+
+                            ByteArrayOutputStream os; // TODO: we can probably use Vert.x / Netty buffering here
+                            if (compressionEnabled) {
+                                os = new ByteArrayOutputStream(contentLength);
+                                clientRequest.putHeader("Content-Encoding", "gzip");
```
Comment (brunobat) [inline extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxHttpExporter.java:135]: this needs to read the `quarkus.otel.exporter.otlp.traces.compression` property

**Review by brunobat — CHANGES_REQUESTED** [review]

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java (hunk 1 above)
Comment (geoand) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java:96]: I was following the suggestion by @radcortez [here](https://github.com/quarkusio/quarkus/issues/21535#issuecomment-1658218695). Let me know which one you guys want, and I'll update the PR

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxHttpExporter.java (hunk 3 above)
Comment (geoand) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxHttpExporter.java:135]: The value of that is passed to the ctor of `VertxHttpExporter` by `OtlpRecorder` (same as for the gRPC exporter). 
Furthermore, the tests already cover this

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java (hunk 2 above)
Comment (geoand) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java:97]: Again, I'll let you and @radcortez decide :)

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java (hunk 1 above)
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java:96]: The spec doesn't define the specific string to use, as far as I can see. That's an implementation prerogative and they chose the values from the link above.

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxHttpExporter.java (hunk 3 above)
Comment (brunobat) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/VertxHttpExporter.java:135]: ok

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java (hunk 1 above)
Comment (geoand) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java]: In any case, I updated the PR

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java (hunk 2 above)
Comment (geoand) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java]: I removed the JSON field completely

File: extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java (hunk 1 above)
Comment (radcortez) [reply extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/config/runtime/exporter/OtlpExporterTracesConfig.java]: It's fine, use `http/protobuf`.

**Review by brunobat — APPROVED** [review]
Looks good to me. Thanks @geoand !

**Comment by geoand** [comment]
🙏🏼

**Comment by rsvoboda** [comment]
@geoand Could you pls open PR for docs?

**Comment by geoand** [comment]
@rsvoboda not sure what you are asking for

**Comment by rsvoboda** [comment]
This is new feature/enhancement, bu this PR didn't add anything into docs/guides.
Could you extend otel related guide(s) with information about OTLP/HTTP option?

**Comment by geoand** [comment]
That's probably best done by @brunobat who has all the OpenTelemetry context

**Comment by geoand** [comment]
That's probably best done by @brunobat who has all the OpenTelemetry context

**Comment by rsvoboda** [comment]
Thanks Bruno. We have a note, should be good for now. It can be extended if people need more details.

**Pull request merged**